src/libmongoc/tests/unified/runner.c

```diff
@@ -20,89 +20,729 @@
+         }
+         BSON_FOREACH_END;
+
+         if (!mongoc_bulk_operation_execute (bulk_insert, NULL, error)) {
+            goto loopexit;
+         }
+      } else {
+         /* Test does not need data inserted, just create the collection. */
+         db = mongoc_client_get_database (test_runner->internal_client, database_name);
+         if (!mongoc_database_create_collection (db, collection_name, wc_opts, error)) {
+            goto loopexit;
+         }
+      }
+
+      ret = true;
+
+   loopexit:
+      mongoc_bulk_operation_destroy (bulk_insert);
+      bson_destroy (wc_opts);
+      bson_destroy (drop_opts);
+      bson_destroy (documents);
+      mongoc_write_concern_destroy (wc);
+      mongoc_collection_destroy (coll);
+      if (!ret) {
+         return false;
+      }
+   }
+   BSON_FOREACH_END;
+   return true;
+}
+
+/* This returns an error on failure instead of asserting where possible.
+ * This allows the test runner to perform server clean up even on failure (e.g.
+ * disable failpoints).
+ */
+bool
+test_run (test_t *test, bson_error_t *error)
+{
+   test_runner_t *test_runner;
+   test_file_t *test_file;
+
+   test_file = test->test_file;
+   test_runner = test_file->test_runner;
+
+   if (test->skip_reason != NULL) {
+      MONGOC_DEBUG ("SKIPPING test '%s'. Reason: '%s'",
+                    test->description,
+                    test->skip_reason);
+      return true;
+   }
+
+   if (test->run_on_requirements) {
+      const char *reason;
+      if (!check_run_on_requirements (
+             test_runner, test->run_on_requirements, &reason)) {
+         MONGOC_DEBUG ("SKIPPING test '%s'. Reason: '%s'",
+                       test->description,
+                       test->skip_reason);
+         return true;
+      }
+   }
+
+   if (!test_setup_initial_data (test, error)) {
+      return false;
+   }
+
+
+   return true;
 }
 
 void
-run_one_test (bson_t *test)
+run_one_test_file (bson_t *bson)
 {
-   test_ctx_init (test);
-   MONGOC_DEBUG ("running test: %s", test_ctx.description);
-   signal (SIGABRT, handle_abort);
-   check_schema_version ();
-   ASSERT (bson_has_field (test, "runOnRequirements"));
-   signal (SIGABRT, SIG_DFL);
-   test_ctx_cleanup ();
+   test_runner_t *test_runner;
+   test_file_t *test_file;
+
+   test_diagnostics_init ();
+
+   test_runner = test_runner_new ();
+   test_runner_terminate_open_transactions (test_runner);
+   test_file = test_file_new (test_runner, bson);
+
+   MONGOC_DEBUG ("running test file: %s", test_file->description);
+
+   check_schema_version (test_file);
+   if (test_file->run_on_requirements) {
+      const char *reason;
+      if (!check_run_on_requirements (
+             test_runner, test_file->run_on_requirements, &reason)) {
+         MONGOC_DEBUG ("SKIPPING test file (%s). Reason:\n%s",
+                       test_file->description,
+                       reason);
+         goto done;
+      }
+   }
```

Comment:
I had it that way originally, but individual tests can have additional run on requirements. L680 checks the run on requirements for an individual test. So I wanted `check_run_on_requirements` to be generic enough to be used in both cases.

That is fair that the caller would need to explicitly check, but I think this will be the only caller for a test file's run on requirements, and L680 will be the only caller for a test's run on requirements. And if the test was not skipped, you'd be informed when the test failed when running against an incompatible set of requirements. So I'm not too concerned.